libiocage/lib/DevfsRules.py

```diff
@@ -0,0 +1,381 @@
+        """
+        Create a ruleset from the provided filters
+
+        Args:
+
+            rule_name (string):
+                Name of the rule that get's created
+
+            rule_number (int):
+                Number of the rule that get's created
+        """
+
+        ruleset = DevfsRuleset(ruleset_name, ruleset_number)
+        for line in self.active_filters:
+            ruleset.append(line)
+        return ruleset
+
+class DevfsRuleset(list):
+    """
+    Representation of a devfs ruleset in the devfs.rules file
+
+    DevfsRuleset instances behave like standard lists and can store strings
+    that  multiple lines (string)
+    """
+
+    PATTERN = r"""^\[(?P<name>[a-z](?:[a-z0-9\-_]*[a-z0-9])?)=
+                (?P<number>[0-9]+)\]\s*(?:#\s*(?P<comment>.*))?$"""
+
+    def __init__(self, value=None, number=None, comment=None):
+        """
+        Initialize DevfsRuleset
+
+        Args:
+
+            value (string): (optional)
+                If specified in combination with a number, this parameter is
+                interpreted as the ruleset name. Otherwise it is parsed with
+                the expectation to find a [<name>=<number>] ruleset definition.
+
+                When value is not specified or None, DevfsRuleset is assumed
+                to be new or unspecified (cannot be exported before name and
+                number were assigned at a later time)
+
+            number (int): (optional)
+                The number of the ruleset. Must be specified to export the
+                ruleset, but is like the name not required to compare the 
+                ruleset with another
+        """
+
+        # when only one argument is passed, it's a line that need to be parsed
+        if value is None and number is None:
+            # name and number will be assigned later
+            name = None
+        elif number is None:
+            name, number, comment = self._parse_line(value)
+        else:
+            name = int(value)
+
+        self.name = name
+        self.number = number
+        self.comment = comment
+        list.__init__(self)
+
+    def has_rule(self, rule):
+        """
+        Returns true if the rule is part of the current ruleset
+
+        Args:
+
+            rule (string):
+                The rule string to be compared with current rules of the
+                ruleset instance
+        """
+        return rule in self
+
+    def append(self, rule):
+        if rule not in self:
+            list.append(self, rule)
+
+    def clone(self, source_ruleset):
+        """
+        Clones the rules from another ruleset
+
+        Args:
+
+            source_ruleset (libiocage.lib.DevfsRules.DevfsRuleset):
+                Ruleset to copy all rules from
+        """
+        for rule in source_ruleset:
+            self.append(rule)
+
+    def _parse_line(self, line):
+
+        # marks beginning of a new ruleset
+        ruleset_match = re.search(DevfsRuleset.PATTERN, line)
```

Comment:
how come you don't have to compile the pattern to tell it that it's `/x`?

Reply:
I do 🙈 ... fixed